### PR TITLE
Ajax image upload

### PR DESCRIPTION
**Pull request by boltnev**
Ajax image uploading and cropping;
Also item form buttons made \<button\> instead of \<a\> (except the case of get-button)

### CODE REVIEW

File: iktomi/cms/ajax_file_upload.py [hunk 1]
```diff
@@ -112,26 +112,27 @@ def _collect_related_fields(self, env, form_field, image,
...
-                    'source_size': image.size,
-                    'original_name': original_name,
-                    })
+                rel_images.append({rel_form_field.input_name:{
+                    "current_url":env.file_manager.get_transient_url(rel_transient, env),
+                    "transient_name":rel_transient.name,
+                    "original_name": original_name,
+                    "mode":"transient",
+                    "source_url":source.url,
+                    "source_transient":source.name,
```
Comment (Harut) [inline iktomi/cms/ajax_file_upload.py]: Mark as hack please! The cllient should have everything to find the source field itself

File: iktomi/cms/ajax_file_upload.py [hunk 2]
```diff
@@ -112,26 +112,27 @@ def _collect_related_fields(self, env, form_field, image,
...
-                rel_images.append({
-                    "name": rel_form_field.input_name,
-                    "file": rel_transient.name,
-                    'file_url': env.file_manager.get_transient_url(rel_transient, env),
-                    'fill_from': form_field.input_name,
-                    'transformations': transforms,
-                    'source_size': image.size,
-                    'original_name': original_name,
-                    })
+                rel_images.append({rel_form_field.input_name:{
```
Comment (Harut) [inline iktomi/cms/ajax_file_upload.py]: this will not work for nested fields

File: iktomi/cms/forms/files.py [hunk 3]
```diff
@@ -28,31 +28,23 @@ def __init__(self, *args, **kwargs):
             kwargs['conv'] = conv(required=required)
         FileFieldSet.__init__(self, *args, **kwargs)
 
-    def get_diff(field1, field2):
```
Comment (Harut) [inline iktomi/cms/forms/files.py]: is this functionality fixed or just dropped off?

Comment (boltnev) [inline iktomi/cms/forms/files.py]: this functionality is dropped off for a while

Comment (Harut) [inline iktomi/cms/forms/files.py]: if you brake something, please do not remove all the evidences :)  Even more, make everything for broken place to be noticed by anyone and fixed in future

File: iktomi/cms/forms/widgets.py [hunk 4]
```diff
@@ -305,7 +305,7 @@ class AjaxFileInput(FileInput):
     def js_config(self):
         return {'input': self.id,
                 'error': self.field.error,
-                'upload_url': self.upload_url,
+                'upload_url': self.upload_url or self.field.upload_url,
```
Comment (Harut) [inline iktomi/cms/forms/widgets.py]: what is the difference between `widget.upload_url` and `field.upload_url`? I don't like the same property is present in two classes. May be, determine the right place and drop it from another?

File: iktomi/cms/static/js/widgets/fieldset.js [hunk 5]
```diff
@@ -113,6 +113,10 @@
             var data = this.getPropsData();
             for (var i=0; i<this.props.widgets.length; i++){
                 var prop = _clone(this.props.widgets[i]);
+                if(!prop.key){
+                    // Auto key property setting     
+                    prop.key = "auto-key-" + i;
+                }
```
Comment (Harut) [inline iktomi/cms/static/js/widgets/fieldset.js]: A you sure this is correct? Once created key must be persistent in whole form lifecycle. Is is guaranteed?

File: iktomi/cms/static/js/widgets/fileinput.js [hunk 6]
```diff
@@ -88,35 +101,31 @@ Widgets.AjaxFileInput = Widgets.create(Widgets.Widget, {
...
-                                     key="cancel-button"
-                                     className="progress_cancel" 
-                                     onClick={this.cancelUploading}>Отмена</a>
-                                     <Widgets.ProgressBar complete={this.state.progress}/>
-                                </div>;
-        }
+        return urlDiv;
+    },
+    inputElements: function(){
+        return [<input type="file"
```
Comment (Harut) [inline iktomi/cms/static/js/widgets/fileinput.js]: Why dont you want to put this just in an render() withou any arrays and keys?

Comment (Harut) [inline iktomi/cms/static/js/widgets/fileinput.js]: render: function() {
         return <div>	
                  {this.urlDiv()}
                  {this.progressContainer()}
                  <input ..>
                  <input ..>
               </div>;

Comment (Harut) [inline iktomi/cms/static/js/widgets/fileinput.js]: Ok! I see you use it in ImageInput

**Pull request merged**